README
README

 BiESAT --> Bi207 Energy Spectra Analysis Toolkit
==================================================

WARNING! this macro has been tested using ROOT versions
'5.14.00e' to '5.24.00' so there is a possibility of 
things breaking when using other versions of ROOT.

for example fitting do the following in ROOT:
root[0] .x BiESAT.cxx("run195")

assuming you have done a 'cvs checkout BiESAT' this
should work, ie the fit should go through all interations
and produce a final result of the fit.

the .conf file is of most importance to the user.
first create a rootfile from your pedistal, beta,
and gamma run data files, in ROOT:

root[0] .x data2root.cxx("ped-file","beta-file","gamma-file",chan)

chan --> either 1 or 9, 1 for lowres and 9 for hihes using VME
         with VME DAQ, lowres --> 0-7 and hires --> 8-15
         but chan 0 is damaged meaning chan 8 is also damaged

the generated files will be placed in ./DBtemp because the user
must first edit the .conf file and confirm the generated files
are correct. the user MUST edit the generated .conf file for the
correct 400keV compton edge position and 976keV CE peak position
in ADC bins (see exampls in ./DBruns). some other settings must
be adjusted to help BiESAT understand what it is trying to fit.

describe --> a string of information describing your setup
             this is useful for when other people want a quick
             overview of your setup and this way you don't have
             to go rumaging through your logbook.

pedrun   --> is the pedistal run file name
betarun  --> is the beta+gamma run file name
gammarun --> is the gamma run file name

gamma    --> is the 400keV compton edge position in ADC bins
beta     --> is the 976keV peak position in ADC bins
Eloss    --> is the estimated energy loss of the electrons
             through the air and reflective material

lowres   --> if the resolution is worse than ~8-9% set this
             to '1' because then BiESAT changes the fitting
             procedure a bit (too much to explain here).
thick    --> if your scint is thicker than ~5cm set this to
             '1' because BiESAT changes the fitting procedure
             a bit (too much to explain here).
fitopt   --> add any ROOT fitting options here you might want
             to test out (from TH1::Fit).
             http://root.cern.ch/root/html/TH1.html#TH1:Fit
rebin    --> sets the rebin for the histograms
maxbin   --> the maximum bin in your histograms (4000 if VME,
             2000 if CAMAC etc.)

once you have all the correct settins, copy over the files
into the ./DBruns directory. the files should include:
name.C, name.conf, name.png, name.root

then in ROOT do:

root[0] .x BiESAT("name")

and hopfully everything will work.


update
------------------
if you have files in the ./DBsimu folder, then do:
root[0] .x BiESAT("name",1)

and if you have files in the ./DBtemp folder, then do:
root[0] .x BiESAT("name",2)




Changelog
Changelog

 BiESAT --> Bi207 Energy Spectra Analysis Toolkit
==================================================

12.10.25
+ plotting the individual K, L, and M shell gaussians
  for the 976keV peak.

09.07.17
~ updated the 09.05.29 changes because the version on
  my laptop was more recent than the version on
  my pc at UCL.

09.07.07
~ officially now called BiESAT.cxx
~ tidy up the macro a bit

09.06.04
~ change first sigma aprox. to be 15% of adc position
~ change lower bounds on 976keV of the 1st fit

09.05.29
~ changed around many fitting parameters to fit lowres better
~ changed how calib is calculated from the gamma fit
+ add subtraction fit at the end of the program

09.05.26
+ root > 5.17 now works
~ changed the way calib is calculated for lowres and thick
~ changed around some fitting settings

09.05.22
+ delete func; to end of macro, seems to stop crashing
+ shifting the gamma hist to match beta at 500 compton
~ let exp0 free for first beta fit

08.11.03
+ use data2root.cxx to convert to rootfile
- NO MORE SUPPORT FOR RAW DATA FILES!!!

08.10.31
~ changed char[]'s to *char[]'s
~ fixed a bug when sped < s482 
+ combined in the rootfile method of reading data

08.10.22
~ changed the way the hists from rootfile are readin
~ changed offset calculatin to use 976-position and global calib
~ fixed the problem with scaling the histos!

08.09.30
+ tweaked parLimits for fitting gamma and multi-gamma
+ fixed the fit to account for multi scattering comptons
+ fixed parameters and if statements for thick and lowres options
~ changed gamma fit to use full fit

08.09.18
+ added condition to make sure compt1000 is less than gaus1000

08.09.09
+ added 'thick' option that deals with multi compton scattering
+ added in a section that normalizes the gamma spectra

08.09.08
- removed previous 'subtracted the Eloss contribution' NOT VALID
+ subtracted the Eloss contribution to sigma from measurement

08.09.05
- removed the gamma energy loss factor



_combineMC.cxx
gROOT->Reset();

#include "./include/manip_files.hxx"
#include "./include/manip_histos.hxx"

///////////////////////////////////////////////////////////
const Int_t input_type = 1;  // 0==rootfile : 1==datafile
const Int_t bins       = 4000;
const Int_t histmax    = 4000;
///////////////////////////////////////////////////////////

Int_t _combineMC(string fname="combined", string outdir="./DBtemp/")
{
  beautify();
  
  if(input_type==0){
    char betafile[256]="MCbetas/results.root";
    char gammafile[256]="MCgammas/results.root";
    char hname[24]="h100";
    
    TFile *file1=new TFile(betafile,"READ");
    TH1F *MCbeta=(TH1F*)file1->Get(hname);
    MCbeta->SetName("MCbeta");
    MCbeta->SetTitle("MCbeta");
    
    TFile *file2=new TFile(gammafile,"READ");
    TH1F *MCgamma=(TH1F*)file2->Get(hname);
    MCgamma->SetName("MCgamma");
    MCgamma->SetTitle("MCgamma");
  }
  
  if(input_type==1){
    char betafile[256]="MCbetas/output.dat";
    char gammafile[256]="MCgammas/output.dat";
    
    TH1F *MCbeta = readfile(betafile,0,0,2,"MCbeta",1,histmax);
    TH1F *MCgamma = readfile(gammafile,0,0,2,"MCgamma",1,histmax);
    MCbeta->SetTitle("MCbeta");
    MCgamma->SetTitle("MCgamma");
  }
  
  TCanvas *look=new TCanvas("look","look",0,0,800,600);
  look->Divide(1,2,.02,.02,0);
  look->cd(1);
  MCbeta->Draw();
  look->Update();
  look->cd(2);
  MCgamma->Draw();
  look->Update();
  
  TString rootname=outdir+fname;
  string s1=rootname+".png";
  string s2=rootname+".C";
  look->Print(s1.c_str());
  look->Print(s2.c_str());
  
  TFile *outfile=gROOT->FindObject(rootname+".root");
  if(outfile) outfile->Close();
  outfile=new TFile(rootname+".root","RECREATE");
  MCbeta->Write();
  MCgamma->Write();
  outfile->Close();
  
  if(genConf(fname,outdir)) return 1;
  
  cout<<"\n\n\tDONE CREATING ==> "<<rootname+".root"<<" \n\n\n";

  return 0;
}


//--- GENERATE A GENERAL CONFIG FILE FOR THE RUNS
Int_t genConf(string fname="combined", string outdir="./DBtemp/"){
  ofstream fout;
  fout.open((outdir+fname+".conf").c_str());
  
  fout<<"void config(){ \n\n";
  
  fout<<"// example = 8in-Ham-SBA-1900v__25x22x10-hex-ej200-esr-mylar__bi-center-6cm \n";
  fout<<"// example = 8in-ETL-1500v_LG-pol-tef__14x14x2-bc404-esr__bi-center-6cm \n\n";
  
  //fout<<"     describe  = \"pmt-info__scint-info__source-info\"; \n\n";
  fout<<"     describe  = \""<<fname.c_str()<<"\"; \n\n";
  
  fout<<"     pedrun    = \"\";          // pedestal run-name \n";
  fout<<"     betarun   = \"MCbeta\";    // beta+gamma run-name \n";
  fout<<"     gammarun  = \"MCgamma\";   // gamma run-name \n\n";
  
  fout<<"     gamma     = 400;     // the 400 kev COMPTON edge position \n";
  fout<<"     beta      = 976;     // the 976 kev CONVERSION-E peak position \n";
  fout<<"     Eloss     = 17.0;    // energy loss in air and reflector \n\n";
  
  fout<<"     lowres    = 0;       // if resolution is >8% set this true \n";
  fout<<"     thick     = 1;       // if using thick scint, set this true \n";
  fout<<"     fitopt    = \"\";      // global fit options (chi^2 or likelyhood etc) \n";
  fout<<"     rebin     = 2;       // rebinning of the histrograms \n";
  fout<<"     maxbin    = "<<histmax<<";    // maximum range you want in your histos \n\n";
  
  fout<<"} \n";
  
  fout.close();  

  return 0;
}

